From dbbb7937f8849f6a0a13c96f57be4d300c73c020 Mon Sep 17 00:00:00 2001
From: Marek Vasut <marex@nabladev.com>
Date: Wed, 8 Oct 2025 13:41:17 +0200
Subject: [PATCH 9001/9001] FIXME: arm64: dts: st: Add TC9595 into STM32MP25xx
 DHCOS SoM and DHSBC

Add support for DSI-to-(e)DP bridge to DHCOS DHSBC.
Currently, the DP-to-USB-C mux is set to fixed orientation
due to missing UCPD support on STM32MP25xx.

This still requires special boot sequence which selects the
correct LTDC clock and which enables supply to the USB-C port:

=> mw 0x44235000 2 ; i2c dev 0 ; i2c mw 0x20 6 0xbf ; i2c mw 0x34 0 0x1c

Signed-off-by: Marek Vasut <marex@nabladev.com>
---
 .../boot/dts/st/stm32mp255c-dhcos-dhsbc.dts   | 78 +++++++++++++++++++
 .../boot/dts/st/stm32mp25xx-dhcos-som.dtsi    | 29 +++++++
 2 files changed, 107 insertions(+)

diff --git a/arch/arm64/boot/dts/st/stm32mp255c-dhcos-dhsbc.dts b/arch/arm64/boot/dts/st/stm32mp255c-dhcos-dhsbc.dts
index 6b07a78069353..ec007c4e10ca3 100644
--- a/arch/arm64/boot/dts/st/stm32mp255c-dhcos-dhsbc.dts
+++ b/arch/arm64/boot/dts/st/stm32mp255c-dhcos-dhsbc.dts
@@ -138,6 +138,84 @@ &i2c2 {
 	status = "okay";
 };
 
+&i2c8 {
+	tc_bridge: bridge@f {
+		compatible = "toshiba,tc9595", "toshiba,tc358767";
+		pinctrl-names = "default";
+		pinctrl-0 = <&mco1_pins_a>;
+		reg = <0xf>;
+		clock-names = "ref";
+		clocks = <&rcc CK_MCO1>;
+		reset-gpios = <&gpiof 14 GPIO_ACTIVE_HIGH>;
+
+		interrupt-parent = <&gpiog>;
+		interrupts = <14 IRQ_TYPE_EDGE_RISING>;
+		toshiba,hpd-pin = <0>;
+
+		ports {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			port@0 {
+				reg = <0>;
+
+				tc_bridge_in: endpoint {
+					data-lanes = <1 2 3 4>;
+					remote-endpoint = <&dsi_out>;
+				};
+			};
+
+			port@2 {
+				reg = <2>;
+
+				endpoint {
+//					toshiba,pre-emphasis = /bits/ 8 <1 1>;
+				};
+			};
+		};
+	};
+};
+
+&dsi {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	status = "okay";
+
+	ports {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		port@0 {
+			reg = <0>;
+			dsi_in: endpoint {
+				remote-endpoint = <&ltdc_ep1_out>;
+			};
+		};
+
+		port@1 {
+			reg = <1>;
+			dsi_out: endpoint {
+				remote-endpoint = <&tc_bridge_in>;
+				data-lanes = <1 2 3 4>;
+			};
+		};
+	};
+};
+
+&ltdc {
+	status = "okay";
+
+	port {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		ltdc_ep1_out: endpoint@0 {
+			reg = <0>;
+			remote-endpoint = <&dsi_in>;
+		};
+	};
+};
+
 &sdmmc1 {
 	pinctrl-names = "default", "opendrain", "sleep";
 	pinctrl-0 = <&sdmmc1_b4_pins_b>;
diff --git a/arch/arm64/boot/dts/st/stm32mp25xx-dhcos-som.dtsi b/arch/arm64/boot/dts/st/stm32mp25xx-dhcos-som.dtsi
index d8b9ef708e214..605d2dc784061 100644
--- a/arch/arm64/boot/dts/st/stm32mp25xx-dhcos-som.dtsi
+++ b/arch/arm64/boot/dts/st/stm32mp25xx-dhcos-som.dtsi
@@ -249,6 +249,13 @@ &gpioa {
 			  "DHCOS-E", "DHCOS-J", "", "",
 			  "DHCOS-D", "", "", "",
 			  "", "", "", "";
+
+	hpd-hog {
+		gpio-hog;
+		gpios = <4 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "E-HPDIN";
+	};
 };
 
 &gpiof {
@@ -277,6 +284,13 @@ &gpioi {
 			  "", "", "", "",
 			  "", "DHCOS-C", "", "",
 			  "", "", "", "";
+
+	hpd-hog {	// CTL1
+		gpio-hog;
+		gpios = <9 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "C-HPDIN_CTL1";
+	};
 };
 
 &gpioz {
@@ -314,6 +328,21 @@ ioexp: gpio@20 {
 			"#PCIe_RST_1_2", "#PCIe_Wake_P1_3",
 			"#QSPI_RST_P1_4", "#CSI0_PWDN_P1_5",
 			"#CSI0_RST_P1_6", "#ETH3_RST_P1_7";
+
+// i2c dev 0 ; i2c mw 0x20 6 0xbf ; i2c mw 0x34 0 0x1c
+		usb-pwren-hog {
+			gpio-hog;
+			gpios = <6 GPIO_ACTIVE_HIGH>;
+			output-high;
+			line-name = "PWREN";
+		};
+
+		hpd-hog {	// FLIP
+			gpio-hog;
+			gpios = <9 GPIO_ACTIVE_HIGH>;
+			output-high; // low means POL on the cable must be pulled LOW with 330R
+			line-name = "FLIP";
+		};
 	};
 
 	tmp110: temperature-sensor@40 {
-- 
2.53.0

